django/db/migrations/state.py

```diff
@@ -95,10 +95,13 @@ def __init__(self, models=None, real_apps=None):
         self.is_delayed = False
         # {remote_model_key: {model_key: [(field_name, field)]}}
         self.relations = None
+        self.resolve_fields_and_relations()
```

Comment:
I don't think we should automatically resolve fields and relations at initialization time as they won't be needed most of the time.

I'd suggest using the same approach as the `apps` `cached_property` where methods are in charge of looking up `'relations' in self.__dict__` to determine if the cache needs updating or not.